tensorizer/serialization.py

```diff
@@ -4446,17 +4128,492 @@ def write_state_dict(self, state_dict: Union[Dict, List, Tuple]):
+                clone_tasks.append(
+                    self._computation_pool.submit(
+                        self._do_clone, w, clone_dependencies
+                    )
+                )
+                w.user_owns_tensor_data = False
+
+            shared_write_specs[0].tensor_data_task = _FutureGroup(
+                clone_tasks + clone_dependencies.futures
+            )
+            for w in shared_write_specs[1:]:
+                w.tensor_data_task = _FutureGroup(clone_tasks)
+
+        for w in write_specs:
+            w.include_crc32 = False
+
+            if w.data_length == 0:
+                # All headers are expected to have crypt_info segments, so add
+                # an empty one
+                w.crypt_info = _crypt_info.CryptInfo()
+                continue
+
+            if w.tensor_data_task is not None:
+                w.tensor_data_task.result(_TIMEOUT)
+                w.tensor_data_task = None
+
+            tensor_memory: memoryview = w.tensor_memoryview
+            chunked = _Chunked(
+                total_size=tensor_memory.nbytes,
+                chunk_size=self._crypt_chunk_size,
+            )
+            nonces = self._new_nonces(chunked.count)
+            w.encryptor = _crypt.ChunkedEncryption(
+                key=self._encryption.key,
+                buffer=tensor_memory,
+                chunk_size=self._crypt_chunk_size,
+                nonces=nonces,
+                executor=self._computation_pool,
+            )
+
+            key_derivation_chunk = self._encryption._crypt_info_chunk()
+            encryption_algorithm_chunk = _crypt_info.XSalsa20ParallelChunk(
+                chunk_size=self._crypt_chunk_size,
+                nonce=nonces[0],
+                macs=w.encryptor.macs,
+            )
+            chunks: Sequence[Any]
+            if key_derivation_chunk is not None:
+                chunks = (key_derivation_chunk, encryption_algorithm_chunk)
+            else:
+                chunks = (encryption_algorithm_chunk,)
+            w.crypt_info = _crypt_info.CryptInfo(chunks)
+
+    def _prepare_for_write_headers(
+        self, write_specs: Sequence[_WriteSpec]
+    ) -> None:
+        # We first need to construct the headers so that we know the size of each
+        for w in write_specs:
+            dtype_bytes = w.dtype.encode("utf-8")  # type: ignore
+            if len(dtype_bytes) >= 256:
+                raise ValueError("dtype name length should be less than 256")
+
+            w.header = _TensorHeaderSerializer(
+                w.module_index,
+                w.tensor_type,
+                w.name.serialized_(),  # name as bytes
+                dtype_bytes,
+                w.shape,
+                w.data_length,
+                0,  # bogus file_offset. This gets filled in in build()
+                include_crc32=w.include_crc32,
+                include_sha256=w.include_sha256,
+                crypt_info=w.crypt_info,
+            )
```

Comment:
The comment `# bogus file_offset. This gets filled in in build()` isn't accurate; it just gets filled in a bit later in this function but before `build()`. It's the tensor data offset that gets filled in in `build()`.
```suggestion
            w.header = _TensorHeaderSerializer(
                w.module_index,
                w.tensor_type,
                w.name.serialized_(),  # name as bytes
                dtype_bytes,
                w.shape,
                w.data_length,
                0,  # placeholder file_offset
                include_crc32=w.include_crc32,
                include_sha256=w.include_sha256,
                crypt_info=w.crypt_info,
            )
```